### PR TITLE
Jwt auditing

### PR DESCRIPTION
**Pull request by ben851**
# Summary | Résumé

First crack at JWT event auditing.

## Related Issues | Cartes liées

*  https://app.zenhub.com/workspaces/notify-planning-core-6411dfb7c95fb80014e0cab0/issues/gh/cds-snc/notification-planning-core/912

# Test instructions | Instructions pour tester la modification

Do JWT stuff and validate the logs appear correctly

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [ ] This PR does not break existing functionality.
- [ ] This PR does not violate GCNotify's privacy policies.
- [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: app/authentication/auth.py [hunk 1]
```diff
@@ -231,13 +239,28 @@ def requires_auth():
...
+        audit_jwt_event("jwt.validated", auth_token, service_id=service.id, api_key_id=api_key.id)
         return
     else:
         # service has API keys, but none matching the one the user provided
+        if failure_reason == "invalid_signature":
+            current_app.logger.warning(
+                "Rejected JWT with invalid signature for service %s, client %s",
+                service.id,
+                request.headers.get("User-Agent"),
+            )
```
Comment (jimleroyer) [inline app/authentication/auth.py:262]: Why add that extra logging rather than pack it into its own `audit_jwt_event` function call?

File: app/authentication/auth.py [hunk 2]
```diff
@@ -187,37 +192,40 @@ def requires_auth():
     try:
         service = dao_fetch_service_by_id_with_api_keys(client)
     except DataError:
+        audit_jwt_event("jwt.authorization_failed", auth_token, reason="invalid_service_id")
```
Comment (jimleroyer) [inline app/authentication/auth.py:195]: Can we verify which function gets through for users leveraging the client API ([gov.uk libraries](https://documentation.notification.canada.ca/en/clients.html))? I assume it's that one, but we need to make sure the client libraries are covered by the new audit (in this PR or a different one).

File: tests/app/authentication/test_audit.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+from notifications_python_client.authentication import create_jwt_token
+from notifications_utils import logging as notify_logging
+from notifications_utils import request_helper
+
+from app.authentication import auth
+from app.authentication.audit import audit_jwt_event
+
+
+@pytest.fixture
+def audit_app(caplog):
```
Comment (jimleroyer) [inline tests/app/authentication/test_audit.py:19]: oh interesting, `caplog`, to capture the log, I didn't see that in pytest before

**Review by jimleroyer — APPROVED** [review]
LGTM - I had a few questions though to clear out. I will approve as this does fit our requirements and it's fairly clean.

Recommendations for additional test coverage (AI generated but I can explain):

> Add an explicit check to ensure you're testing JWT-specific authentication. You could:
>
> 1. Assert the auth type in your test fixture — mock get_auth_token to verify it returns JWT_AUTH_TYPE
> 2. Add a negative test — verify that ApiKey-v1 headers do NOT trigger JWT auditing

Especially the last one would be nice to test the edge conditions (we shouldn't audit non jwt usage).

**Review by jzbahrai — APPROVED** [review]
also looked at this and makes sense to me.

File: app/authentication/auth.py (hunk 1 above)
Comment (ben851) [reply app/authentication/auth.py:262]: Agreed, I removed it.

**Comment by ben851** [comment]
> LGTM - I had a few questions though to clear out. I will approve as this does fit our requirements and it's fairly clean.
> 
> Recommendations for additional test coverage (AI generated but I can explain):
> 
> > Add an explicit check to ensure you're testing JWT-specific authentication. You could:
> > 
> > 1. Assert the auth type in your test fixture — mock get_auth_token to verify it returns JWT_AUTH_TYPE
> > 2. Add a negative test — verify that ApiKey-v1 headers do NOT trigger JWT auditing
> 
> Especially the last one would be nice to test the edge conditions (we shouldn't audit non jwt usage).

I've added those tests and addressed your other comments

File: app/authentication/auth.py (hunk 2 above)
Comment (ben851) [reply app/authentication/auth.py:195]: Yes it is, and I added a test that explicitly uses these libraries.
